### PR TITLE
Fixes #29939 - drop category column from settings

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Settings table was holding category column to support old style of  settings, but it is now ready to be decomissioned as plugins were given  enough time.

I'll try to take this PR to the merge, but if you have an idea, feel free to push to my branch! :)

### CODE REVIEW

**Comment by ezr-ondrej** [comment]
No idea if the test failures are related :(

**Comment by ekohl** [comment]
No, https://github.com/theforeman/foreman/pull/9531 was intended to fix the tests.

**Comment by ezr-ondrej** [comment]
> No, #9531 was intended to fix the tests.

Right, there is one more failure, but now I at least know it is not related :)

File: test/fixtures/settings.yml
Comment (ekohl) [inline test/fixtures/settings.yml:1]: Can we merge https://github.com/theforeman/foreman/pull/9530 first? Then we simply get rid of the file altogether.

**Review by ekohl — COMMENTED** [review]
I do wonder about plugins and migrations. We do have some plugin migrations where we explicitly change the category. If we install that plugin after this migration ran then I think it'll fail.

**Comment by ezr-ondrej** [comment]
> We do have some plugin migrations where we explicitly change the category. If we install that plugin after this migration ran then I think it'll fail.

Yeah, we will have failures in plugins unfortunatelly, but should be easy to find and fix. I don't know when is the next release, but there is no reason to rush this, so timing it right and give plugins enough time is probably a good idea :+1:

File: test/fixtures/settings.yml
Comment (ezr-ondrej) [reply test/fixtures/settings.yml:1]: Sure :shrug: I don't think the order matters much here tho, if #9530 gets in first, I'm happy to rebase, but I don't think it should be a blocker :)

**Comment by stejskalleos** [comment]
I did quick test with following plugins:
* Tasks
* Rex
* Ansible
* Google
* katello

Foreman booted without any problems and basic scenarios worked as expected,
except the Ansible, but that's not a problem of this PR.
That one failed unit test is issue from libvirt, not related to these changes.
Since 3.5 branching is behind us I'm for merging this ASAP as it will give
us more time to catch all the issues before the next branching.


@ekohl Do you want to get https://github.com/theforeman/foreman/pull/9530 first (there are some related unit-failures) or this one first?

**Comment by stejskalleos** [comment]
Waiting for https://github.com/theforeman/foreman/pull/9530, failed test is not related

**Comment by ezr-ondrej** [comment]
> Waiting for #9530, failed test is not related

I wouldn't see #9530 as hard blocker as it's easy to rebase after this is merged. I'm not against waiting, but it sounded like we have good time window opportunity and there is not much activity in the other PR :)

Tho I'm happy to wait for another opportunity, this had waited long time already ^_^

**Comment by stejskalleos** [comment]
@ezr-ondrej Yeah you are right, let's get this in. Can you please rebase, there are some conflicts.

**Comment by ezr-ondrej** [comment]
Sure, rebased! :)

**Comment by ezr-ondrej** [comment]
And it's :green_apple: now :)

**Review by stejskalleos — APPROVED** [review]
:green_apple: **LGTM**,

Foreman works and running REX/Ansible jobs went well too.
@ekohl any other comments? From my side we can merge the PR

**Comment by ekohl** [comment]
My only concern is plugin breakage. I don't have time to deal with any fallout, but that shouldn't hold this back. If you feel confident it can move forward (since it is a good direction) and look into any failures, please go ahead.

**Comment by stejskalleos** [comment]
Thanks @ezr-ondrej @ekohl !

**Comment by evgeni** [comment]
> My only concern is plugin breakage. I don't have time to deal with any fallout, but that shouldn't hold this back. If you feel confident it can move forward (since it is a good direction) and look into any failures, please go ahead.

Well, this did now break installation of every single plugin that had a migration of the kind of 
```ruby
Setting.where(category: 'Setting::Plugin').update_all(category: 'Setting')
```
which is *every* plugin that had a setting and got that migration to migrate from the `Setting::Plugin` to the `Setting` category.

Here is an example of me trying to install the latest Foreman Tasks on an empty nightly:

```
== 20211123170430 TasksSettingsToDslCategory: migrating =======================
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::UndefinedColumn: ERROR:  column settings.category does not exist
LINE 1: ... IN (SELECT "settings"."id" FROM "settings" WHERE "settings"...
                                                             ^
…
/usr/share/foreman/vendor/ruby/2.7.0/gems/foreman-tasks-7.2.0/db/migrate/20211123170430_tasks_settings_to_dsl_category.rb:3:in `up'
```

TTBOMK this doesn't happen when you install plugins *directly* with the foreman installation, as then the plugin-category-change migration happens before the foreman-category-drop migration and this "hides" the issue in our current pipelines.

**Comment by evgeni** [comment]
If I get it right, every single of those migrations needs to be updated to read:
```ruby
Setting.where(category: 'Setting::Plugin').update_all(category: 'Setting') if column_exists?(:settings, :category)
```

**Comment by ezr-ondrej** [comment]
Oh cool, I've just noticed this and @stejskalleos has that fixed already :exploding_head:

**Comment by stejskalleos** [comment]
Yop:
* https://github.com/theforeman/foreman-tasks/pull/705
* https://github.com/theforeman/foreman_remote_execution/pull/778
* https://github.com/theforeman/foreman_ansible/pull/589
* https://github.com/theforeman/foreman_bootdisk/pull/140
* https://github.com/theforeman/foreman_discovery/pull/595
* https://github.com/theforeman/foreman_monitoring/pull/88
* https://github.com/theforeman/foreman_host_extra_validator/pull/12
* https://github.com/theforeman/foreman_expire_hosts/pull/71
* https://github.com/theforeman/foreman_salt/pull/180
* https://github.com/theforeman/puppetdb_foreman/pull/80
* https://github.com/theforeman/foreman_rh_cloud/pull/797
* https://github.com/theforeman/foreman_templates/pull/167
* https://github.com/Katello/katello/pull/10432

**Comment by evgeni** [comment]
I think we're still missing releases for most of these gems, and thus 3.6+ is broken right now.

**Pull request merged**
